### PR TITLE
Add memory-efficient merging script

### PR DESCRIPTION
**Pull request by airaria**
## Description

This PR adds a memory-efficient merging script (merge_llama_with_chinese_lora_low_mem.py, and related notebooks).
Comparing to the original script, the new script significantly reduces the memory usage for merging LoRAs to the LLaMA model. 

Estimated max memory usage when merging alpaca-lora to the LLaMA with the new script:

| Model | Max Memory Usage |
| ----- | ---------------- |
| 7B    | 15GB             |
| 13B   | 18GB             |
| 33B   | 22GB             |

### CODE REVIEW

File: scripts/merge_llama_with_chinese_lora_low_mem.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+                    state_dict[k] += (
+                        transpose(
+                            tandl['state_dict'][lora_key_B].float() 
+                          @ tandl['state_dict'][lora_key_A].float(), tandl['fan_in_fan_out']) * tandl['scaling']
+                    )
+            weight_size = state_dict[k].numel() * dtype_byte_size(state_dict[k].dtype)
+            total_size += weight_size
+
+        if output_type=='huggingface':
+            print(f"Saving ckpt {filename} to {output_dir} in HF format...")
```
Comment (iMountTai) [inline scripts/merge_llama_with_chinese_lora_low_mem.py:312]: Converting to huggingface format...

File: scripts/merge_llama_with_chinese_lora_low_mem.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+            print("Cannot find lora model on the disk. Downloading lora model from hub...")
+            lora_model_path = snapshot_download(repo_id=lora_model_path)
+        tokenizer = LlamaTokenizer.from_pretrained(lora_model_path)
+        lora_config = peft.LoraConfig.from_pretrained(lora_model_path)
+        lora_state_dict = torch.load(os.path.join(lora_model_path,'adapter_model.bin'),map_location='cpu')
+        if 'base_model.model.model.embed_tokens.weight' in lora_state_dict:
+            lora_vocab_size = lora_state_dict['base_model.model.model.embed_tokens.weight'].shape[0]
+            assert lora_vocab_size==len(tokenizer), \
+            (f"The vocab size of the tokenizer {len(tokenizer)} does not match the vocab size of the LoRA weight {lora_vocab_size}.\n"
+            "Did you misuse the LLaMA tokenizer with the Alpaca-LoRA weight?\n"
```
Comment (iMountTai) [inline scripts/merge_llama_with_chinese_lora_low_mem.py:247]: 会不会出现alpaca tokenzier 与llama-lora混用的情况？

File: scripts/merge_llama_with_chinese_lora_low_mem.py (hunk 1 above)
Comment (airaria) [reply scripts/merge_llama_with_chinese_lora_low_mem.py:312]: > Converting to huggingface format...

The the loaded ckpt is in HF format, there is no "converting" step when saving to HF format

**Review by iMountTai — APPROVED** [review]

**Comment by ymcui** [comment]
GPT-4 with Code Interpreter made some suggestions.
Please make any changes that seem valid and appropriate.

---

Here are some suggestions on how to improve the code:

1. **Avoid Hardcoding Values**: The model parameters are hardcoded in `params_of_models` which may make the script less flexible. Consider loading these parameters from a configuration file or allowing the user to input them as arguments.
2. **Modularize the Code**: The code could be more modular. It's generally a good idea to separate logic into functions, which makes the code easier to read, understand, test, and maintain. For instance, you could separate the loading of models and tokenizers into its own function.
3. **Error Handling**: While the script does contain some error handling logic, more comprehensive error handling could improve robustness. For example, what if the base model doesn't exist even after trying to download it? Or what if the LoRA models cannot be loaded for some reason?
4. **Documentation**: Consider adding more comments to explain what different parts of the code are doing, especially for complex operations. This is especially important in the `save_shards` and `merge_shards` functions, which perform non-trivial manipulations on the model weights.
5. **Naming Conventions**: Some variable names are not clear or descriptive, such as `tandl` and `ti`. Choose descriptive variable names to improve the readability of your code.
6. **Code Duplication**: There are several instances where you perform the same operations multiple times, which could be avoided. For instance, the `split` operation in `save_shards` function is called multiple times with different parameters. You could create a helper function for this operation to avoid code duplication.
7. **Use Context Managers for File Operations**: When handling file operations, it's a good practice to use context managers (`with` statement) to ensure resources are properly managed. For example, in the last part of your code where you are loading and saving JSON files, you should use the `with` statement to open the file.

Overall, the code seems to be doing a complex task (merging LoRA models with a base model) and appears to be doing it correctly, but these improvements could make the code cleaner, more efficient, and easier to maintain.

**Review by iMountTai — APPROVED** [review]

**Pull request merged**